Ubuntu 17.10 undeclared X509_V_ERR_INVALID_CALL

rust-openssl does not build with default libssl-1.0.2g shipped with Ubuntu-17.10/16.04.

with latest master (commit eb5fda588f83e74701bbcaeb646ca3800275d495 (HEAD -> master, origin/master, origin/HEAD)) 

the cargo build fails on  Ubuntu 17.10 with following error. 

It seems that Ubuntu 17.10 claims to provide libssl-1.0.2g but in fact it is libssl-1.0.0. The following definitions are missing in header files: X509_V_ERR_INVALID_CALL, X509_V_ERR_STORE_LOOKUP, X509_V_ERR_PROXY_SUBJECT_NAME_VIOLATION

Not sure how to handle this, seems more like a failure of Ubuntu-distribution. Just, it is annoying: even using an up-to-date distro, rust-openssl systests are  failing.

$ dpkg -l | grep libssl
ii  libssl-dev:amd64                           1.0.2g-1ubuntu13.3                          amd64        Secure Sockets Layer toolkit - development files
ii  libssl-doc                                 1.0.2g-1ubuntu13.3                          all          Secure Sockets Layer toolkit - development documentation
ii  libssl1.0.0:amd64                          1.0.2g-1ubuntu13.3                          amd64        Secure Sockets Layer toolkit - shared libraries
```

   Compiling systest v0.1.0 (file:///home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/systest)
error: failed to run custom build command for `systest v0.1.0 (file:///home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/systest)`
process didn't exit successfully: `/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-a8dc8dfeef8948ab/build-script-build` (exit code: 101)
--- stdout
-----------------------------------------
cargo:rerun-if-changed=../openssl-sys/src/lib.rs
cargo:rerun-if-changed=../openssl-sys/src/ossl10x.rs
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
CC_x86_64-unknown-linux-gnu = None
CC_x86_64_unknown_linux_gnu = None
HOST_CC = None
CC = None
HOST = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
CFLAGS_x86_64-unknown-linux-gnu = None
CFLAGS_x86_64_unknown_linux_gnu = None
HOST_CFLAGS = None
CFLAGS = None
DEBUG = Some("true")
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-Wall" "-Wextra" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-Wno-type-limits" "-o" "/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.o" "-c" "/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.c"
cargo:warning=/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.c:2578:67: error: ‘X509_V_ERR_INVALID_CALL’ undeclared here (not in a function); did you mean ‘X509_V_ERR_INVALID_CA’?
cargo:warning=             static int __test_const_X509_V_ERR_INVALID_CALL_val = X509_V_ERR_INVALID_CALL;
cargo:warning=                                                                   ^~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=                                                                   X509_V_ERR_INVALID_CA
cargo:warning=/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.c:2584:67: error: ‘X509_V_ERR_STORE_LOOKUP’ undeclared here (not in a function); did you mean ‘X509_V_ERR_CERT_REVOKED’?
cargo:warning=             static int __test_const_X509_V_ERR_STORE_LOOKUP_val = X509_V_ERR_STORE_LOOKUP;
cargo:warning=                                                                   ^~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=                                                                   X509_V_ERR_CERT_REVOKED
cargo:warning=/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.c:2590:83: error: ‘X509_V_ERR_PROXY_SUBJECT_NAME_VIOLATION’ undeclared here (not in a function); did you mean ‘X509_V_ERR_EXCLUDED_VIOLATION’?
cargo:warning=             static int __test_const_X509_V_ERR_PROXY_SUBJECT_NAME_VIOLATION_val = X509_V_ERR_PROXY_SUBJECT_NAME_VIOLATION;
cargo:warning=                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cargo:warning=                                                                                   X509_V_ERR_EXCLUDED_VIOLATION
exit code: 1

--- stderr
thread 'main' panicked at '

Internal error occurred: Command "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-Wall" "-Wextra" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-Wno-type-limits" "-o" "/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.o" "-c" "/home/frehberg/src/proj-rust-wasm-sign/tmp/rust-openssl/target/debug/build/systest-77c6b0a896b5b988/out/all.c" with args "cc" did not execute successfully (status code exit code: 1).

', /home/frehberg/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.6/src/lib.rs:2043:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,9 +10,10 @@
 * Added `RsaPrivateKeyBuilder` to allow control over initialization of optional components of an RSA
     private key.
 * Added DER encode/decode support to `SslSession`.
-* openssl-sys now provides the `OPENSSL_VERSION_NUMBER` and `OPENSSL_LIBRESSL_VERSION_NUMBER`
-    environment variables to downstream build scripts which contains the hex-encoded version number
-    of the OpenSSL or LibreSSL distribution being built against. The other variables are deprecated.
+* openssl-sys now provides the `DEP_OPENSSL_VERSION_NUMBER` and
+    `DEP_OPENSSL_LIBRESSL_VERSION_NUMBER` environment variables to downstream build scripts which
+    contains the hex-encoded version number of the OpenSSL or LibreSSL distribution being built
+    against. The other variables are deprecated.
 
 ## [v0.10.6] - 2018-03-05
 

--- a/openssl-sys/build.rs
+++ b/openssl-sys/build.rs
@@ -437,6 +437,10 @@ See rust-openssl README for more information:
         let openssl_version = openssl_version.unwrap();
         println!("cargo:version_number={:x}", openssl_version);
 
+        if openssl_version >= 0x1_00_02_08_0 {
+            println!("cargo:rustc-cfg=ossl102h");
+        }
+
         if openssl_version >= 0x1_01_02_00_0 {
             version_error()
         } else if openssl_version >= 0x1_01_01_00_0 {

--- a/openssl-sys/src/ossl10x.rs
+++ b/openssl-sys/src/ossl10x.rs
@@ -1,13 +1,13 @@
+use std::io::{self, Write};
+use std::mem;
+use std::process;
+use std::ptr;
 use std::sync::{Mutex, MutexGuard};
 use std::sync::{Once, ONCE_INIT};
-use std::mem;
-use std::ptr;
-use std::process;
-use std::io::{self, Write};
 
-use libc::{c_char, c_int, c_long, c_uchar, c_uint, c_ulong, c_void, size_t};
 #[cfg(not(ossl101))]
 use libc::time_t;
+use libc::{c_char, c_int, c_long, c_uchar, c_uint, c_ulong, c_void, size_t};
 
 #[repr(C)]
 pub struct stack_st_ASN1_OBJECT {
@@ -131,7 +131,7 @@ pub struct DSA {
 #[repr(C)]
 pub struct ECDSA_SIG {
     pub r: *mut BIGNUM,
-    pub s: *mut BIGNUM
+    pub s: *mut BIGNUM,
 }
 
 #[repr(C)]
@@ -719,8 +719,11 @@ pub const CRYPTO_LOCK_X509: c_int = 3;
 pub const CRYPTO_LOCK_SSL_CTX: c_int = 12;
 pub const CRYPTO_LOCK_SSL_SESSION: c_int = 14;
 
+#[cfg(ossl102h)]
 pub const X509_V_ERR_INVALID_CALL: c_int = 65;
+#[cfg(ossl102h)]
 pub const X509_V_ERR_STORE_LOOKUP: c_int = 66;
+#[cfg(ossl102h)]
 pub const X509_V_ERR_PROXY_SUBJECT_NAME_VIOLATION: c_int = 67;
 
 static mut MUTEXES: *mut Vec<Mutex<()>> = 0 as *mut Vec<Mutex<()>>;

